### PR TITLE
handles lists restore collision

### PR DESCRIPTION
**Pull request by HiteshRepo**
handles lists restore collisions

for collection/site/restore_test, along with tests for lists restore collision, changes for the refactoring of existing test setup also got mingled.

#### Does this PR need a docs update or release note?
- [x] :no_entry: No

#### Type of change

- [x] :sunflower: Feature

#### Issue(s)
#4754 

#### Test Plan

- [x] :muscle: Manual
- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

**Review by ryanfkeepers — COMMENTED** [review]
I'm having a difficult time figuring out which code was moved, deleted, or simply changed.  For example, /site/restore.go's `ConsumeRestoreCollections` is deleted, but I can't tell if that was unused code or if it was rerouted elsewhere, or something else.  I'm going to hold on further review and ask that this gets split up into two PRs (one for movement/deletion, one for changes) to make it easier to review.

**Comment by HiteshRepo** [comment]
> I'm having a difficult time figuring out which code was moved, deleted, or simply changed. For example, /site/restore.go's `ConsumeRestoreCollections` is deleted, but I can't tell if that was unused code or if it was rerouted elsewhere, or something else. I'm going to hold on further review and ask that this gets split up into two PRs (one for movement/deletion, one for changes) to make it easier to review.

@ryanfkeepers Please find the PR that removes unused/non-required code: https://github.com/alcionai/corso/pull/5043

File: src/internal/m365/collection/site/restore_test.go
Comment (HiteshRepo) [reply src/internal/m365/collection/site/restore_test.go:216]: @ryanfkeepers 
No they are not moved from `/sharepoint/restore_test.go`. The test was inconsequential as the core logic is tested in `collection` package.

File: src/internal/m365/collection/site/restore.go
Comment (ryanfkeepers) [inline src/internal/m365/collection/site/restore.go:139]: I'm confused by what's going on with the behavior here.  As a general statement, extra comments are always helpful for other maintainers so that we can understand what's going on in a process.  This is one of those places where a little explanation would go a long way.

So, what I figure is happening is that the func does this:

1. look up the current list
2. delete the current list
3. restore the backed-up list
4. if no error occurs when restoring, good, we exit.
5. if an error occurs when restoring, try to restore the list that we just deleted.

I think I have that right.  If so: that works, but it's not my favorite pattern, because it contains a lot more moving parts and points of failure.  I think you could slim it down to the following for a simpler flow and fewer points of failure.  At least, I think this is possible, assuming the newly created list has a different ID than the previous one.

1. Restore the backed-up list.  If an error occurs, return.
2. Delete the current list.  If an error occurs, return.

And if that works, we might not need this extra func at all.  Up on lines ~85-105 you could do this instead:
```go
 restoredList, err = rh.PostList(ctx, newName, storedList, errs)
 if err != nil {
 	return dii, clues.WrapWC(ctx, err, "restoring list")
 }

if collisionPolicy != control.Replace {
  rh.DeletetList(...)
}
```

File: src/internal/m365/collection/site/restore.go
Comment (ryanfkeepers) [inline src/internal/m365/collection/site/restore.go:80]:
```suggestion
			return dii, clues.Stack(core.ErrAlreadyExists)
```

File: src/internal/m365/collection/site/restore_test.go
Comment (ryanfkeepers) [inline src/internal/m365/collection/site/restore_test.go:254]: various nits:

Naming structure doesn't follow standards.  It should be `Test<FunctionUnderTest>_<variation>`
```suggestion
func (suite *SharePointRestoreSuite) TestRestoreListItem_collisions() {
```

All four new tests duplicate a lot of boilerplate declarations and behavior.  This bloat makes the tests more difficult to read and, in the long term, will require more effort to maintain and refactor.  They should be condensed into a single test using a test table.

File: src/internal/m365/collection/site/restore_test.go
Comment (ryanfkeepers) [inline src/internal/m365/collection/site/restore_test.go:508]: nit: prefer `stub` to suggest we're mocking out a list.  `create` could be confused with the process of adding it to an external dataset.
```suggestion
func stubList(listTemplate, listDisplayName string) models.Listable {
```

**Review by ryanfkeepers — CHANGES_REQUESTED** [review]
The general behavior looks reasonable, but I'm going to hold off on approval because I think the implementation could use some hefty re-work.  There are a couple choices here with the potential for problems that we'd want to avoid.

File: src/internal/m365/collection/site/restore.go
Comment (HiteshRepo) [reply src/internal/m365/collection/site/restore.go:127]: @ryanfkeepers 
for `replace` collision, we need to create list with same name. It is not possible to have 2 lists with same name. Hence following the below order:
- Fetch the collided list and store it temporarily
- Delete collided list
- Create stored list
- If Create stored list fails, reinstate the collided list

File: src/internal/m365/collection/site/restore.go
Comment (HiteshRepo) [reply src/internal/m365/collection/site/restore.go:139]: @ryanfkeepers 
you got it right. I would have gone for the pattern you mentioned as well, but in `replace` collision, we need to create backed-ip list with same name as current/collided list. But site does not allow to create two lists with same name.

File: src/internal/m365/collection/site/restore.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/site/restore.go:127]: hmm, what do you think about, instead:
1. restore list with a name like `corso_restore_<date>_<originalName>
2. delete old list
3. patch restored list name to `originalName`

It's still fewer steps/calls, and avoids the potential for full data loss.

File: src/internal/m365/collection/site/restore.go
Comment (ryanfkeepers) [reply src/internal/m365/collection/site/restore.go:139]: Thanks for confirming!  I grok the need to work around it.  I pitched a third workflow option in the other comment; let me know what you think.

File: src/internal/m365/collection/site/restore.go
Comment (HiteshRepo) [reply src/internal/m365/collection/site/restore.go:127]: @ryanfkeepers 
This change request has been implemneted in below 3 PRs [approved]:
- https://github.com/alcionai/corso/pull/5072
- https://github.com/alcionai/corso/pull/5073
- https://github.com/alcionai/corso/pull/5074

File: src/internal/m365/collection/site/restore_test.go
Comment (HiteshRepo) [reply src/internal/m365/collection/site/restore_test.go:254]: @ryanfkeepers 
this change has been accommodated in https://github.com/alcionai/corso/pull/5074

**Comment by HiteshRepo** [comment]
@ryanfkeepers moving ahead with merger of this, as change requests for this PR has been addressed and approved in other PRs:
- https://github.com/alcionai/corso/pull/5072
- https://github.com/alcionai/corso/pull/5073
- https://github.com/alcionai/corso/pull/5074

**Pull request merged**
